streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -270,17 +276,36 @@ private void maybeAddNodeForOptimizationMetadata(final GraphNode node) {
 
     // use this method for testing only
     public void buildAndOptimizeTopology() {
-        buildAndOptimizeTopology(false);
+        buildAndOptimizeTopology(null);
     }
 
-    public void buildAndOptimizeTopology(final boolean optimizeTopology) {
+    public void buildAndOptimizeTopology(final Properties props) {
+        // Vicky: Do we need to verify props?
+        final List<String> optimizationConfigs;
+        if (props == null) {
+            optimizationConfigs = new ArrayList<>();
+            optimizationConfigs.add(StreamsConfig.NO_OPTIMIZATION);
+        } else {
+            final StreamsConfig config = new StreamsConfig(props);
+            optimizationConfigs = config.getList(StreamsConfig.TOPOLOGY_OPTIMIZATION_CONFIG);
+        }
 
         mergeDuplicateSourceNodes();
-        if (optimizeTopology) {
-            LOG.debug("Optimizing the Kafka Streams graph for repartition nodes");
+        if (optimizationConfigs.contains(StreamsConfig.OPTIMIZE)
```

Comment:
nit: probably easier to read if we just have one top level branch for `contains(OPTIMIZE)`:
```
if (contains(OPTIMIZE)) {
  applyOptimizations()
}

void applyOptimizations() {
  if (contains(REUSE) reuse
  if (contains(MERGE) merge...
}
```

Reply:
A user can optimize either by providing the config `all` which is the `StreamsConfig.OPTIMIZE` which will apply all optimization rules or by specifying a list of optimization rules. With your suggestion, an optimization will be applied only if `contains(OPTIMIZE)` is true which is not correct. WDYT?